api/error_unexpected_response.go

```diff
@@ -0,0 +1,164 @@
+func (e UnexpectedResponseError) Unwrap() error             { return e.err }
+
+func NewUnexpectedResponseError(src UnexpectedResponseErrorSource, opts ...UnexpectedResponseErrorOption) UnexpectedResponseError {
+	new := src()
+	for _, opt := range opts {
+		opt(new)
+	}
+	if new.statusText == "" {
+		// the stdlib's http.StatusText function is a good place to start
+		new.statusFromCode(http.StatusText)
+	}
+
+	return *new
+}
+
+// Use textual representation of the given integer code. Called when status text
+// is not set using the WithStatusText option.
+func (e UnexpectedResponseError) statusFromCode(f func(int) string) {
+	e.statusText = f(e.statusCode)
+	if !e.HasStatusText() {
+		e.statusText = "unknown status code"
+	}
+}
+
+func (e UnexpectedResponseError) Error() string {
+	var eTxt strings.Builder
+	eTxt.WriteString("Unexpected response code")
+	if e.HasBody() || e.HasStatusCode() {
+		eTxt.WriteString(": ")
+	}
+	if e.HasStatusCode() {
+		eTxt.WriteString(fmt.Sprint(e.statusCode))
+		if e.HasBody() {
+			eTxt.WriteRune(' ')
+		}
+	}
+	if e.HasBody() {
+		eTxt.WriteString(fmt.Sprintf("(%s)", e.body))
+	}
+
+	return eTxt.String()
+}
+
+// UnexpectedResponseErrorOptions are functions passed to NewUnexpectedResponseError
+// to customize the created error.
+type UnexpectedResponseErrorOption func(*UnexpectedResponseError)
+
+// WithError allows the addition of a Go error that may have been encountered
+// while processing the response. For example, if there is an error constructing
+// the gzip reader to process a gzip-encoded response body.
+func WithError(e error) UnexpectedResponseErrorOption {
+	return func(u *UnexpectedResponseError) { u.err = e }
+}
+
+// WithBody overwrites the Body value with the provided custom value
+func WithBody(b string) UnexpectedResponseErrorOption {
+	return func(u *UnexpectedResponseError) { u.body = b }
+}
+
+// WithStatusText overwrites the StatusText value the provided custom value
+func WithStatusText(st string) UnexpectedResponseErrorOption {
+	return func(u *UnexpectedResponseError) { u.statusText = st }
+}
+
+// WithExpectedStatuses provides a list of statuses that the receiving function
+// expected to receive. This can be used by API callers to provide more feedback
+// to end-users.
+func WithExpectedStatuses(s []int) UnexpectedResponseErrorOption {
+	return func(u *UnexpectedResponseError) {
+		u.expected = make([]int, len(s))
+		copy(u.expected, s)
+	}
+}
+
+// UnexpectedResponseErrorSource provides the basis for a NewUnexpectedResponseError.
+type UnexpectedResponseErrorSource func() *UnexpectedResponseError
+
+// FromHTTPResponse read an open HTTP response, drains and closes its body as
+// the data for the UnexpectedResponseError.
+func FromHTTPResponse(resp *http.Response) UnexpectedResponseErrorSource {
+	return func() *UnexpectedResponseError {
+		// collect and close the body
+		var buf bytes.Buffer
+		_, _ = io.Copy(&buf, resp.Body)
+		_ = resp.Body.Close()
```

Comment:
`io.Copy` doesn't return an error on EOF, so if we ignore the error here are we ignoring some underlying problem that would be useful to know? Also, is it safe to close this `resp.Body` twice? We've not done so in the only consumer we have here, but it would be easy for someone to do `defer resp.Body.Close()` in a caller and miss that it's being closed here.

By the time we call this, we're already in an error state and things could be screwy, so maybe we should verify that `resp` and `resp.Body` are non-nil?